### PR TITLE
Adds documentation about dependency resolution

### LINKED ISSUE #775: Documentation on DI in EDC
The Dependency Injection mechanism used in EDC to resolve and inject dependencies in extensions should be documented. Proposed topics to cover in the document:
- Usage of DI in EDC
  - @Requires, @Provides
  - Injection with @Inject
  - Resolving services with `ServiceExtensionContext.getService`
- DI Lifecycle (see https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/637#discussion_r814526780)

### PR DESCRIPTION
**Pull request by paullatzelsperger**
## What this PR changes/adds:

Adds developer documentation about dependency resolution and registration.

## Why it does that

On several occasions we've noticed that the dependency resolution mechanism was used incorrectly, so adding the relevant documentation aims to alleviate that.

## Further notes

Does not include detailed documentation about the `ObjectFactory`, the `Injector` or the `InjectionPointScanner` to keep the scope of the document focused on the high-level concepts.

## Linked Issue(s)

Closes #775 

## Checklist

_not relevant_

### CODE REVIEW

File: docs/developer/dependency_resolution.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+The code base of the Eclipse Dataspace Connector is architected in away that allows for easily extending and swapping
+the core functionality using certain plug-points called _extensions_. One example would be to swap out an in-memory
+implementation for a datastore for one backed by an actual database. In order to achieve that there are several key
+components:
+
+- a service interface, typically located in an SPI module
+- a module providing the implementation, typically located in the `extensions` directory
+- the service registry, i.e. the `ServiceExtensionContext`. Since it is not quite an IoC container, we'll henceforth
+  refer to it as the "context".
+- a hook point into the loading sequence: an extension that that instantiates and registers the implementation class
```
Comment (marcgs) [inline docs/developer/dependency_resolution.md:12]:
```suggestion
- a hook point into the loading sequence: an extension that instantiates and registers the implementation class
```

File: docs/developer/dependency_resolution.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  with the context
+
+## Registering a service implementation
+
+As a general rule the module that provides the implementation also has to register it with
+the `ServiceExtensionContext` ("context" for short). This is done in an accompanying service extension. For example,
```
Comment (marcgs) [inline docs/developer/dependency_resolution.md:18]: nit: A bit redundant, you already mention few lines above _we'll henceforth
  refer to it as the "context"_

File: docs/developer/dependency_resolution.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+#### Only available in `ServiceExtensions` (and why that's not entirely true)
+
+Services can only be injected into `ServiceExtension` objects at this time as they are the main hook points for plugins,
+and they have a clearly defined interface. All subsequent object creation must be done manually using conventional
+mechanisms like constructors or builders.
+
+**Well, actually...**
+
+However, using the `ReflectiveObjectFactory` together with an `Injector` and an `InjectionPointScanner` would
```
Comment (marcgs) [inline docs/developer/dependency_resolution.md:149]: I wonder if it's worth mentioning this at all. Wouldn't it be better to treat all these classes as internal?

Comment (paullatzelsperger) [reply docs/developer/dependency_resolution.md]: Yeah, probably better to remove and amend the document, if and when the injection mechanism gets expanded.

**Review by marcgs — APPROVED** [review]

**Review by juliapampus — APPROVED** [review]

**Pull request merged**